blobs/howto/python/blob-devguide-py/blob-devguide-blobs.py

```diff
@@ -0,0 +1,410 @@
+    # </Snippet_upload_blob_file>
+
+    # <Snippet_upload_blob_tags>
+    def upload_blob_tags(self, blob_service_client: BlobServiceClient, container_name):
+        container_client = blob_service_client.get_container_client(container=container_name)
+        sample_tags = {"Content": "image", "Date": "2022-01-01"}
+        with open(file=os.path.join('filepath', 'filename'), mode="rb") as data:
+            blob_client = container_client.upload_blob(name="sample-blob.txt", data=data, tags=sample_tags)
+    # </Snippet_upload_blob_tags>
+
+    # <Snippet_download_blob_file>
+    def download_blob_to_file(self, blob_service_client: BlobServiceClient, container_name):
+        blob_client = blob_service_client.get_blob_client(container=container_name, blob="sample-blob.txt")
+        with open(file=os.path.join('filepath', 'filename'), mode="wb") as sample_blob:
+            download_stream = blob_client.download_blob()
+            sample_blob.write(download_stream.readall())
+    # </Snippet_download_blob_file>
+
+    # <Snippet_download_blob_chunks>
+    def download_blob_chunks(self, blob_service_client: BlobServiceClient, container_name):
+        blob_client = blob_service_client.get_blob_client(container=container_name, blob="sample-blob.txt")
+
+        # This returns a StorageStreamDownloader
+        stream = blob_client.download_blob()
+        chunk_list = []
+
+        # Read data in chunks to avoid loading all into memory at once
+        for chunk in stream.chunks():
+            # Process your data (anything can be done here - `chunk` is a byte array)
+            chunk_list.append(chunk)
+    # </Snippet_download_blob_chunks>
+
+    # <Snippet_download_blob_stream>
+    def download_blob_to_stream(self, blob_service_client: BlobServiceClient, container_name):
+        blob_client = blob_service_client.get_blob_client(container=container_name, blob="sample-blob.txt")
+
+        # readinto() downloads the blob contents to a stream and returns the number of bytes read
+        stream = io.BytesIO()
+        num_bytes = blob_client.download_blob().readinto(stream)
+        print(f"Number of bytes: {num_bytes}")
+    # </Snippet_download_blob_stream>
+
+    # <Snippet_download_blob_text>
+    def download_blob_to_string(self, blob_service_client: BlobServiceClient, container_name):
+        blob_client = blob_service_client.get_blob_client(container=container_name, blob="sample-blob.txt")
+
+        # content_as_text() downloads the blob contents and decodes as text - default values are shown for parameters
+        blob_text = blob_client.download_blob().content_as_text(max_concurrency=1, encoding='UTF-8')
+        print(f"Blob contents: {blob_text}")
+    # </Snippet_download_blob_text>
+
+    # <Snippet_list_containers>
+    def list_containers(self, blob_service_client: BlobServiceClient):
+        i=0
+        all_pages = blob_service_client.list_containers(include_metadata=True, results_per_page=5).by_page()
```

Comment:
Hmm I'm thinking our most basic `list_containers` sample should actually be like the below `prefix` sample without using `by_page` and `results_per_page`. Messing with the pages is definitely more of an advanced scenario that most people won't need to use.

I'm debating whether showing `by_page` is really even needed though. We could add another sample like `list_containers_pages` or something but again, not sure if that's even needed. I'll leave it up to you, but I do think the most basic sample should not include pages.

Reply:
Good thought. I'll simplify the basic case and consider whether we want an addition sample to show paging.